mlir/lib/Conversion/TosaToMIOpen/TosaToMIOpen.cpp

```diff
@@ -297,148 +281,89 @@ class MatMulConverter final : public OpConversionPattern<tosa::MatMulOp> {
-          bottomOp, [&](::mlir::Diagnostic &diag) {
-            diag << "Bottom transpose not matching";
-          });
-    }
-    auto convOp = dyn_cast<tosa::Conv2DOp>(operands[0].getDefiningOp());
-    if (!(convOp)) {
-      return rewriter.notifyMatchFailure(convOp, [&](::mlir::Diagnostic &diag) {
-        diag << "Second is not tosa::conv2d";
-      });
-    }
-
-    if (!checkNCHW(convOp)) {
-      return rewriter.notifyMatchFailure(convOp, [&](::mlir::Diagnostic &diag) {
-        diag << "Convolution doesn't expect NHWC->NCHW conversion";
-      });
-    }
-
-    auto opr0 = *convOp.getODSOperands(0).begin();
-    auto opr1 = *convOp.getODSOperands(1).begin();
-    auto inputTpOp = dyn_cast<tosa::TransposeOp>(opr0.getDefiningOp());
-    auto filterTpOp = dyn_cast<tosa::TransposeOp>(opr1.getDefiningOp());
-    if (!(inputTpOp) || !(filterTpOp)) {
-      return rewriter.notifyMatchFailure(convOp, [&](::mlir::Diagnostic &diag) {
-        diag << "No transpose found for input/filter";
+    if (dims.size() != 4) {
+      return b.notifyMatchFailure(top, [&](::mlir::Diagnostic &diag) {
+        diag << "Bad constant transpose dims";
       });
     }
-    auto inputTpOpr1 = *inputTpOp.getODSOperands(1).begin();
-    auto filterTpOpr1 = *filterTpOp.getODSOperands(1).begin();
-
-    // Try to get transpose to input
-    auto inTpConstOp = dyn_cast<arith::ConstantOp>(inputTpOpr1.getDefiningOp());
-    if (!inTpConstOp) {
-      inTpConstOp = dyn_cast<arith::ConstantOp>(
-          inputTpOpr1.getDefiningOp()->getOperand(0).getDefiningOp());
-    }
-    SmallVector<int64_t> inTpPerm;
-    auto inTpConstAttr =
-        inTpConstOp->getAttr("value").cast<DenseElementsAttr>();
-    for (auto i : inTpConstAttr.getValues<int64_t>()) {
-      inTpPerm.push_back(i);
-    }
 
-    // Try to get transpose to filter
-    auto filTpConstOp =
-        dyn_cast<arith::ConstantOp>(filterTpOpr1.getDefiningOp());
-    if (!filTpConstOp) {
-      filTpConstOp = dyn_cast<arith::ConstantOp>(
-          filterTpOpr1.getDefiningOp()->getOperand(0).getDefiningOp());
+    Value tInput = top.getOperand(0);
+    Value tOutput = top.getResult();
+
+    if (tosa::Conv2DOp convOp = tInput.getDefiningOp<tosa::Conv2DOp>()) {
+      // tosa.conv2d output is transpose
+      permuteLayout(convOp, "output_layout", "nhwk", dims);
+      convOp->getResult(0).setType(tOutput.getType());
+      top->replaceAllUsesWith(convOp);
+    } else {
+      // trace output to tosa.conv2d
+      for (auto &use : tOutput.getUses()) {
+        if (auto op = dyn_cast<tosa::Conv2DOp>(use.getOwner())) {
+          if (convOp)
+            return failure();
+          convOp = op;
+        } else {
+          return failure();
+        }
+      }
+
+      // conv Input Modifier
+      if (convOp.getOperand(0) == tOutput) {
+        // input feature map
+        permuteLayout(convOp, "input_layout", "nhwc", dims, true);
+        top.replaceAllUsesWith({tInput});
```

Comment:
And per above, for the multi-convolution case etc, you could just replace that particular use.

(and you'd return failure() on the next go-around through the pattern when you find no convolutions that can be rewritten)

Reply:
see comment above.